### PR TITLE
API - Add | Remove Routing Rules 

### PR DESCRIPTION
**Pull request by hossinasaadi**
add rules : `./xray api adrules --server=127.0.0.1:62789 c1.json c2.json`
remove rules : `./xray api rmrules --server=127.0.0.1:62789 ruleTag1 ruleTag2`

**also add ruleTag to identify rules.**

```json
{
...
    "routing": {
...
      "rules": [
        {
          "ruleTag" : "mainRule",
          "outboundTag": "api",
          "type": "field",
          "inboundTag": [
            "api"
          ]
        },
...
```

### CODE REVIEW

**Comment by Fangliding** [comment]
It seems the test has failed

**Comment by vrnobody** [comment]
How about replace the entire routing object? That is easier to implement and use. More over we can hot-switch some balancing strategies too.

**Comment by hossinasaadi** [comment]
> It seems the test has failed

now should be ok

File: app/router/router.go [hunk 1]
```diff
@@ -80,6 +87,80 @@ func (r *Router) PickRoute(ctx routing.Context) (routing.Route, error) {
...
 }
 
+// AddRule implements routing.Router.
+func (r *Router) AddRule(config *serial.TypedMessage) error {
+
+	inst, err := config.GetInstance()
+	if err != nil {
+		return err
+	}
+	if c, ok := inst.(*Config); ok {
```
Comment (yuhan6665) [inline app/router/router.go:97]: I see @vrnobody made a comment that it is useful to replace all existing rules. I think it is a good point that rules relate to each other. Can we make the api accept an array of Configs here, not just one?
Then maybe we can add a flag in AddRuleRequest to control if we want to append or replace.

File: main/commands/all/api/rules_add.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+		conf, err := serial.DecodeJSONConfig(r)
+		if err != nil {
+			base.Fatalf("failed to decode %s: %s", arg, err)
+		}
+		rcs = append(rcs, *conf.RouterConfig)
+	}
+	if len(rcs) == 0 {
+		base.Fatalf("no valid rule found")
+	}
+	for _, in := range rcs {
```
Comment (yuhan6665) [inline main/commands/all/api/rules_add.go:58]: Try to make this change to one call with the list of configs

**Review by yuhan6665 — COMMENTED** [review]
Thanks for your PR! This is an important new feature. I've made some comments. Looking forward to it

File: app/router/router.go [hunk 3]
```diff
@@ -19,6 +21,9 @@ type Router struct {
 	rules          []*Rule
 	balancers      map[string]*Balancer
 	dns            dns.Client
+
+	mu     sync.Mutex
+	config *Config
```
[the remark this answers is not in the document]
Comment (hossinasaadi) [reply app/router/router.go:26]: yes, i use it before and forgot to remove it.

File: app/router/router.go (hunk 1 above)
Comment (hossinasaadi) [reply app/router/router.go:97]: > Can we make the api accept an array of Configs here, not just one?

why array of configs? we currently can pass multiple json configs to AddRule api and each config can includes number of rules as well!. ( let me know if i didn’t understood the point)

> Then maybe we can add a flag in AddRuleRequest to control if we want to append or replace.

i think we can add this flag to api. seems cool

File: main/commands/all/api/rules_add.go (hunk 2 above)
Comment (hossinasaadi) [reply main/commands/all/api/rules_add.go:58]: you mean i should check if any configs entered (length  of unnamedArgs)? 

https://github.com/XTLS/Xray-core/blob/6127251547685707869c9f2b86256239fd539c1e/main/commands/all/api/rules_add.go#L44-L45

**Comment by hossinasaadi** [comment]
> How about replace the entire routing object? That is easier to implement and use. More over we can hot-switch some balancing strategies too.

hot-switch balancer and replace|append new routing configs added.
@yuhan6665

File: app/router/router.go (hunk 1 above)
Comment (yuhan6665) [reply app/router/router.go:101]: My bad, I see now you already pass in the whole rule list

**Comment by yuhan6665** [comment]
Thanks a lot! When I have time, I will integrate https://github.com/XTLS/Xray-core/pull/412 as the next step. (Primarily to support GetRoutingRules API)

**Comment by mzbqhbc** [comment]
I run the command line and get this error
`failed to perform AddRule: rpc error: code = Unimplemented desc = unknown service xray.app.router.command.RoutingService`

**Comment by mzbqhbc** [comment]
I run the command line and get this error
`failed to perform AddRule: rpc error: code = Unimplemented desc = unknown service xray.app.router.command.RoutingService`

**Comment by hossinasaadi** [comment]
> RoutingService



> I run the command line and get this error `failed to perform AddRule: rpc error: code = Unimplemented desc = unknown service xray.app.router.command.RoutingService`

```json
    "api": {
      "services": [
        "RoutingService"
      ]
    },

```

**Comment by mzbqhbc** [comment]
> > RoutingService
> 
> > I run the command line and get this error `failed to perform AddRule: rpc error: code = Unimplemented desc = unknown service xray.app.router.command.RoutingService`
> 
> ```json
>     "api": {
>       "services": [
>         "RoutingService"
>       ]
>     },
> ```

Thanks

**Comment by isqyz** [comment]
当我尝试添加路由，失败了。

**Comment by isqyz** [comment]
你好，谢谢你贡献了代码。我尝试使用命令行添加路由配置，但失败了。代码
```c#
            return File.Exists(configFilePath) &&
                    LaunchXray($"api adi --server={xrayHost} \"{configFilePath}\"", out _) &&
                    LaunchXray($"api ado --server={xrayHost} \"{configFilePath}\"", out _) &&
                    LaunchXray($"api adrules --server={xrayHost} \"{configFilePath}\"", out _);
``` 
如：预期是能达到和不使用动态添加的效果。错误log:2025/06/16 15:09:19.195880 from tcp:127.0.0.1:64952 accepted tcp:xxx.xxx.xxx:443 [10000357in-tag >> 10000263out-tag]
```
	"routing": {
	"rules": [
      {
        "type": "field",
        "inboundTag": [ "10000263in-tag" ],
        "outboundTag": "10000263out-tag"
      }
    ],
	"rules": [
      {
        "type": "field",
        "inboundTag": [ "10000357in-tag" ],
        "outboundTag": "10000357out-tag"
      }
    ]
  }
```

**Comment by AkatsukiNoKag** [comment]
> > 路由服务
> 
> > 我运行命令行并收到此错误`failed to perform AddRule: rpc error: code = Unimplemented desc = unknown service xray.app.router.command.RoutingService`
> 
> ```json
>     "api": {
>       "services": [
>         "RoutingService"
>       ]
>     },
> ```

I added it on both the server and the client, but it still doesn’t work.

**Pull request merged**